Bio/Seq.py

```diff
@@ -1981,15 +1981,55 @@ def __init__(self, data, length=None):
         if data is None:
             if length is None:
                 raise ValueError("length must not be None if data is None")
-            self._data = _UndefinedSequenceData(length)
+            elif length == 0:
+                self._data = b""
+            elif length < 0:
+                raise ValueError("length must not be negative.")
+            else:
+                self._data = _UndefinedSequenceData(length)
         elif isinstance(data, (bytes, SequenceDataAbstractBaseClass)):
             self._data = data
         elif isinstance(data, (bytearray, _SeqAbstractBaseClass)):
             self._data = bytes(data)
         elif isinstance(data, str):
             self._data = bytes(data, encoding="ASCII")
         elif isinstance(data, dict):
-            self._data = _PartiallyDefinedSequenceData(length, data)
+            if length is None:
+                raise ValueError("length must not be None if data is a dictionary")
+            elif length == 0:
+                self._data = b""
+            elif length < 0:
+                raise ValueError("length must not be negative.")
+            else:
+                end = -1
+                starts = sorted(data.keys())
+                _data = {}
+                for start in starts:
+                    seq = data[start]
+                    if isinstance(seq, str):
+                        seq = bytes(seq, encoding="ASCII")
+                    else:
+                        try:
+                            seq = bytes(seq)
+                        except Exception:
+                            raise ValueError("Expected bytes-like objects or strings")
+                    if start < end:
+                        raise ValueError("Sequence data are overlapping.")
+                    elif start == end:
+                        _data[current] += seq  # noqa: F821
```

Comment:
I think I understand why F821 is raised, ``current`` may not be defined - for example if ``starts == [-1]`` although that should not happen.

Would it be clearer to add ``current = None`` before the loop? That would trigger a ``TypeError`` if used for slicing which seems better than a ``NameError``.

Reply:
The only possibility for `current` not to be defined is if there is a bug somewhere else in Biopython. Adding `current = None` may mask such a bug.

Reply:
I don't think so - that's why I'm suggesting an initial ``None`` value. The variable ``current`` is local to this method, and only ever used for slicing - so if it does not get properly initialised and remains ``None``, we get a ``TypeError``.

Reply:
I'm OK with a `NameError`.